src/mlpack/core/data/split_data.hpp

```diff
@@ -179,50 +155,43 @@ void StratifiedSplit(const arma::Mat<T>& input,
  * @param shuffleData If true, the sample order is shuffled; otherwise, each
  *       sample is visited in linear order. (Default true.)
  */
-template<typename T, typename U>
+template<typename T, typename LabelsType,
+         typename = std::enable_if_t<arma::is_arma_type<LabelsType>::value> >
 void Split(const arma::Mat<T>& input,
-           const arma::Row<U>& inputLabel,
+           const LabelsType& inputLabel,
            arma::Mat<T>& trainData,
            arma::Mat<T>& testData,
-           arma::Row<U>& trainLabel,
-           arma::Row<U>& testLabel,
+           LabelsType& trainLabel,
+           LabelsType& testLabel,
            const double testRatio,
            const bool shuffleData = true)
 {
   const size_t testSize = static_cast<size_t>(input.n_cols * testRatio);
   const size_t trainSize = input.n_cols - testSize;
   trainData.set_size(input.n_rows, trainSize);
   testData.set_size(input.n_rows, testSize);
-  trainLabel.set_size(trainSize);
-  testLabel.set_size(testSize);
 
+  arma::uvec order = arma::linspace<arma::uvec>(0, input.n_cols - 1,
+      input.n_cols);
   if (shuffleData)
+    order = arma::shuffle(order);
+
+  if (trainSize > 0)
   {
-    arma::uvec order = arma::shuffle(arma::linspace<arma::uvec>(
-        0, input.n_cols - 1, input.n_cols));
-    if (trainSize > 0)
-    {
-      trainData = input.cols(order.subvec(0, trainSize - 1));
-      trainLabel = inputLabel.cols(order.subvec(0, trainSize - 1));
-    }
-    if (trainSize < input.n_cols)
-    {
-    testData = input.cols(order.subvec(trainSize, input.n_cols - 1));
-    testLabel = inputLabel.cols(order.subvec(trainSize, input.n_cols - 1));
-    }
+    trainLabel.set_size(1, trainSize);
+    trainData = input.cols(order.subvec(0, trainSize - 1));
+
+    for (size_t i = 0; i < trainSize; ++i)
+      trainLabel(0, i) = inputLabel(0, order(i));
```

Comment:
Looks like you can use the same strategy as for `trainData` here as well, and relay on `cols` and `subvec` to avoid the for loop.

Reply:
Another note about using `order` when `shuffleData` is `false` is that the compiler has no way to reason that `order` is just `[1, 2, ..., n]`.  That in turn means that it can't use SIMD instructions to copy multiple elements in each instruction.  That actually can make a noticeable runtime difference. :+1: